codelists/views.py

```diff
@@ -26,30 +38,87 @@ def index(request):
     return render(request, "codelists/index.html", ctx)
 
 
-@login_required
-def create_codelist(request, project_slug):
-    project = get_object_or_404(Project, slug=project_slug)
-    if request.method == "POST":
-        form = CodelistForm(request.POST, request.FILES)
-        if form.is_valid():
-            codelist = actions.create_codelist(
-                project=project,
-                name=form.cleaned_data["name"],
-                coding_system_id=form.cleaned_data["coding_system_id"],
-                description=form.cleaned_data["description"],
-                methodology=form.cleaned_data["methodology"],
-                csv_data=form.cleaned_data["csv_data"].read().decode("utf8"),
-            )
-            return redirect(codelist)
-    else:
-        form = CodelistForm()
+@method_decorator(login_required, name="dispatch")
+class CreateCodelist(TemplateView):
+    ReferenceFormSet = formset_factory(ReferenceForm)
+    SignOffFormSet = formset_factory(SignOffForm)
+    template_name = "codelists/create_codelist.html"
+
+    @transaction.atomic()
+    def all_valid(self, form, reference_formset, signoff_formset):
+        codelist = actions.create_codelist(
+            project=self.project,
+            name=form.cleaned_data["name"],
+            coding_system_id=form.cleaned_data["coding_system_id"],
+            description=form.cleaned_data["description"],
+            methodology=form.cleaned_data["methodology"],
+            csv_data=form.cleaned_data["csv_data"].read().decode("utf8"),
+        )
 
-    ctx = {
-        "project": project,
-        "form": form,
-    }
+        # get changed forms so we ignore empty form instances
+        reference_forms = (f for f in reference_formset if f.has_changed())
+        signoff_forms = (f for f in signoff_formset if f.has_changed())
+
+        for form in itertools.chain(reference_forms, signoff_forms):
+            instance = form.save(commit=False)
+            instance.codelist = codelist
+            instance.save()
```

Comment:
Would it be possible to add `references` and `signoffs` as arguments to  `actions.create_codelist()`?

We've now got transaction logic in two places, and if we wanted a factory to create a codelist with references, we'd have to duplicate what we have here.

Reply:
Yep, that makes sense, fixed in 0e203a1